pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java

```diff
@@ -0,0 +1,116 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pulsar.broker.service;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.util.concurrent.FastThreadLocal;
+
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.bookkeeper.mledger.Entry;
+import org.apache.bookkeeper.mledger.impl.PositionImpl;
+import org.apache.pulsar.common.api.Commands;
+import org.apache.pulsar.common.api.proto.PulsarApi.CommandAck.AckType;
+import org.apache.pulsar.common.api.proto.PulsarApi.MessageMetadata;
+
+public abstract class AbstractBaseDispatcher {
+
+    protected final Subscription subscription;
+
+    protected AbstractBaseDispatcher(Subscription subscription) {
+        this.subscription = subscription;
+    }
+
+    /**
+     * Filter messages that are being sent to a consumers.
+     * <p>
+     * Messages can be filtered out for multiple reasons:
+     * <ul>
+     * <li>Checksum or metadata corrupted
+     * <li>Message is an internal marker
+     * <li>Message is not meant to be delivered immediately
+     * </ul>
+     *
+     * @param entries
+     *            a list of entries as read from storage
+     *
+     * @param batchSizes
+     *            an array where the batch size for each entry (the number of messages within an entry) is stored. This
+     *            array needs to be of at least the same size as the entries list
+     *
+     * @param sendMessageInfo
+     *            an object where the total size in messages and bytes will be returned back to the caller
+     * @param subscription
+     *            the subscription object
+     */
+    public void filterEntriesForConsumer(List<Entry> entries, int[] batchSizes, SendMessageInfo sendMessageInfo) {
+        checkArgument(batchSizes.length >= entries.size());
+        int totalMessages = 0;
+        long totalBytes = 0;
+
+        for (int i = 0, entriesSize = entries.size(); i < entriesSize; i++) {
+            Entry entry = entries.get(i);
+            ByteBuf metadataAndPayload = entry.getDataBuffer();
+            PositionImpl pos = (PositionImpl) entry.getPosition();
+
+            MessageMetadata msgMetadata = Commands.peekMessageMetadata(metadataAndPayload, subscription.toString(), -1);
+
+            try {
+                if (msgMetadata == null) {
+                    // Message metadata was corrupted
+                    entries.set(i, null);
+                    entry.release();
+                    subscription.acknowledgeMessage(Collections.singletonList(pos), AckType.Individual,
```

Comment:
add debug log or may be info log because it shouldn't happen often.

Reply:
This is already logged in the metadata parsing method

Reply:
Do we even need to do this on subscription? The operation is to tell the managed ledger to not give us this message again. dispatchers have a handle on the managed ledger.

Reply:
Most likely, yes, though technically it's only true for PersistentDispatcher. This asbtract method it's shared across all the types of dispatchers, hence it goes though the Subscription abstraction.